lib/SILOptimizer/Utils/OptimizerStatsUtils.cpp

```diff
@@ -0,0 +1,970 @@
+/// collected statistics during the compilation. But if there is a need for a
+/// more complex analysis of collected data (e.g. aggregation by a pipeline
+/// stage or by the type of a transformation), it is often better to dump
+/// as much data as possible into a file using e.g. `-sil-stats-dump-all
+/// -sil-stats-modules -sil-stats-functions` and then e.g. use the helper scripts
+/// to store the collected data into a database and then perform complex queries
+/// on it. Many kinds of analysis can be then formulated pretty easily as SQL
+/// queries.
+///
+/// The output format is a set of CSV (comma separated values) lines. Each lines
+/// represents one counter or one counter change.
+///
+/// For counter updates it looks like this:
+///  Kind, CounterName, StageName, TransformName, TransformPassNumber,
+///  DeltaValue, OldCounterValue, NewCounterValue, Duration, Symbol
+///
+/// For counter updates it looks like this:
+///  Kind, CounterName, StageName, TransformName, TransformPassNumber,
+///  CounterValue, Duration, Symbol
+///
+/// where Kind is one of "function", "module", "function_history",
+///       CounterName is one of "block", "inst", "function", "memory",
+///       Symbol is e.g. the name of a function.
+///       StageName and TransformName are the names of the current optimizer
+///       pipeline stage and current transform.
+///       Duration is the duration of the transformation.
+//===----------------------------------------------------------------------===//
+
+#include <iostream>
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Process.h"
+#include "swift/SIL/SILValue.h"
+#include "swift/SIL/SILVisitor.h"
+#include "swift/SILOptimizer/Analysis/Analysis.h"
+#include "swift/SILOptimizer/PassManager/PassManager.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/OptimizerStatsUtils.h"
+using namespace swift;
+
+namespace {
+
+/// The total number of different kinds of SILInstructions.
+constexpr int SILInstructionsNum = int(ValueKind::Last_SILInstruction) + 1;
+
+/// A set of counters, one per SILInstruction kind.
+typedef SmallVector<int, SILInstructionsNum> InstructionCounts;
+
+/// Map a SILInstruction name to its SILInstructionKind.
+SILInstructionKind getSILInstructionKind(StringRef InstName) {
+#define INST(Id, Parent, TextualName, MemoryBehavior, ReleasingBehavior)       \
+  if (InstName == #TextualName)                                                \
+    return SILInstructionKind::Id;
+
+#include "swift/SIL/SILNodes.def"
+
+#ifdef NDEBUG
+  std::cerr
+      << "Unknown insruction name used by the -sil-stats-only-instructions\n";
+  abort();
+#endif
+  llvm_unreachable(
+      "Unknown insruction name used by the -sil-stats-only-instructions");
+}
+
+/// Map SILInstructionKind to a corresponding SILInstruction name.
+const char *getSILInstructionName(SILInstructionKind Kind) {
```

Comment:
Don't we already have this?

Reply:
I couldn't find it somehow. We use something for SIL printing, but there it is tightly integrated with the printer.

Reply:
Seems like something that would generally be useful, why not create a helper for ValueKind. Then you can just case to that and use that.

Reply:
This sounds like a good idea! I'll push the commit doing it very soon.